File: just-talk-app/native/VoiceModule/RtsVoiceModule.m
#import "RtsVoiceModule.h"
#import <Speech/Speech.h>
#import <AVFoundation/AVFoundation.h>

@implementation RtsVoiceModule {
  AVAudioEngine *_audioEngine;
  SFSpeechAudioBufferRecognitionRequest *_recognitionRequest;
  SFSpeechRecognitionTask *_recognitionTask;
  SFSpeechRecognizer *_speechRecognizer;
  BOOL _isMuted;
  BOOL _isStopping;
  long long _sessionStartMs;
}

RCT_EXPORT_MODULE()

+ (BOOL)requiresMainQueueSetup { return NO; }

- (NSArray<NSString *> *)supportedEvents {
  return @[@"onTranscriptEvent", @"onVoiceStateChange", @"onVoiceError"];
}

RCT_EXPORT_METHOD(requestPermissions:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  [SFSpeechRecognizer requestAuthorization:^(SFSpeechRecognizerAuthorizationStatus authStatus) {
    [[AVAudioSession sharedInstance] requestRecordPermission:^(BOOL granted) {
      dispatch_async(dispatch_get_main_queue(), ^{
        if (authStatus == SFSpeechRecognizerAuthorizationStatusAuthorized && granted) {
          resolve(@"granted");
        } else if (authStatus == SFSpeechRecognizerAuthorizationStatusDenied ||
                   authStatus == SFSpeechRecognizerAuthorizationStatusRestricted) {
          resolve(@"denied");
        } else {
          resolve(@"undetermined");
        }
      });
    }];
  }];
}

RCT_EXPORT_METHOD(startListening:(NSString *)language
                  customVocabulary:(NSArray<NSString *> *)customVocabulary
                  resolver:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  NSLocale *locale = [NSLocale localeWithLocaleIdentifier:language];
  _speechRecognizer = [[SFSpeechRecognizer alloc] initWithLocale:locale];

  if (!_speechRecognizer || !_speechRecognizer.isAvailable) {
    reject(@"unavailable", [NSString stringWithFormat:@"Speech recognizer not available for locale %@", language], nil);
    return;
  }

  NSError *error = nil;
  AVAudioSession *session = [AVAudioSession sharedInstance];
  [session setCategory:AVAudioSessionCategoryPlayAndRecord
                  mode:AVAudioSessionModeMeasurement
               options:AVAudioSessionCategoryOptionDefaultToSpeaker |
                       AVAudioSessionCategoryOptionMixWithOthers
                 error:&error];
  if (error) { reject(@"audio_error", error.localizedDescription, error); return; }

  [session setActive:YES withOptions:AVAudioSessionSetActiveOptionNotifyOthersOnDeactivation error:&error];
  if (error) { reject(@"audio_error", error.localizedDescription, error); return; }

  _audioEngine = [[AVAudioEngine alloc] init];
  _recognitionRequest = [[SFSpeechAudioBufferRecognitionRequest alloc] init];
  _recognitionRequest.shouldReportPartialResults = YES;
  if (customVocabulary.count > 0) _recognitionRequest.contextualStrings = customVocabulary;

  _sessionStartMs = (long long)([[NSDate date] timeIntervalSince1970] * 1000);
  _isMuted = NO;
  _isStopping = NO;

  __weak typeof(self) weakSelf = self;
  _recognitionTask = [_speechRecognizer recognitionTaskWithRequest:_recognitionRequest
    resultHandler:^(SFSpeechRecognitionResult *result, NSError *taskError) {
      __strong typeof(weakSelf) strongSelf = weakSelf;
      if (!strongSelf) return;
      if (taskError) {
        if (!strongSelf->_isStopping) {
          [strongSelf sendEventWithName:@"onVoiceError" body:taskError.localizedDescription];
        }
        return;
      }
      if (!result) return;

      long long nowMs = (long long)([[NSDate date] timeIntervalSince1970] * 1000);
      long long latency = nowMs - strongSelf->_sessionStartMs;
      NSString *text = result.bestTranscription.formattedString;
      NSString *stability = result.isFinal ? @"final" : @"partial";

      id confidence = [NSNull null];
      if (result.isFinal && result.bestTranscription.segments.lastObject) {
        confidence = @(result.bestTranscription.segments.lastObject.confidence);
      }

      [strongSelf sendEventWithName:@"onTranscriptEvent" body:@{
        @"text": text,
        @"stability": stability,
        @"timestamp_ms": @(nowMs),
        @"latency_ms_from_audio_start": @(latency),
        @"confidence": confidence,
        @"engine_metadata": @{@"engine": @"apple_speech_transcriber", @"is_final": @(result.isFinal)}
      }];
    }];

  AVAudioInputNode *inputNode = _audioEngine.inputNode;
  AVAudioFormat *format = [inputNode outputFormatForBus:0];
  [inputNode installTapOnBus:0 bufferSize:1024 format:format block:^(AVAudioPCMBuffer *buf, AVAudioTime *when) {
    __strong typeof(weakSelf) strongSelf = weakSelf;
    if (!strongSelf || strongSelf->_isMuted) return;
    [strongSelf->_recognitionRequest appendAudioPCMBuffer:buf];
  }];

  [_audioEngine prepare];
  NSError *startError = nil;
  [_audioEngine startAndReturnError:&startError];
  if (startError) { reject(@"start_error", startError.localizedDescription, startError); return; }

  [self sendEventWithName:@"onVoiceStateChange" body:@"listening"];
  resolve(nil);
}

RCT_EXPORT_METHOD(stopListening:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
  _isStopping = YES;
  [_audioEngine stop];
  [_audioEngine.inputNode removeTapOnBus:0];
  [_recognitionRequest endAudio];
  [_recognitionTask cancel];
  _recognitionRequest = nil;
  _recognitionTask = nil;
  [self sendEventWithName:@"onVoiceStateChange" body:@"stopped"];
  resolve(nil);
}

RCT_EXPORT_METHOD(mute) {
  _isMuted = YES;
  [self sendEventWithName:@"onVoiceStateChange" body:@"muted"];
}

RCT_EXPORT_METHOD(unmute) {
  _isMuted = NO;
  [self sendEventWithName:@"onVoiceStateChange" body:@"listening"];
}

@end

File: road-to-sale-app/ios/VoiceModule/RtsVoiceModule.swift
import Foundation
import Speech
import AVFoundation

@objc(RtsVoiceModule)
class RtsVoiceModule: RCTEventEmitter {

  private var audioEngine = AVAudioEngine()
  private var recognitionRequest: SFSpeechAudioBufferRecognitionRequest?
  private var recognitionTask: SFSpeechRecognitionTask?
  private var speechRecognizer: SFSpeechRecognizer?
  private var isMuted = false
  private var sessionStartMs: Int64 = 0

  override static func requiresMainQueueSetup() -> Bool { return false }

  override func supportedEvents() -> [String]! {
    return ["onTranscriptEvent", "onVoiceStateChange", "onVoiceError"]
  }

  // MARK: - Permissions

  @objc func requestPermissions(_ resolve: @escaping RCTPromiseResolveBlock,
                                 rejecter reject: @escaping RCTPromiseRejectBlock) {
    SFSpeechRecognizer.requestAuthorization { authStatus in
      AVAudioSession.sharedInstance().requestRecordPermission { granted in
        DispatchQueue.main.async {
          switch authStatus {
          case .authorized where granted:
            resolve("granted")
          case .denied, .restricted:
            resolve("denied")
          default:
            resolve("undetermined")
          }
        }
      }
    }
  }

  // MARK: - Start/Stop

  @objc func startListening(_ language: String,
                              customVocabulary: [String],
                              resolver resolve: @escaping RCTPromiseResolveBlock,
                              rejecter reject: @escaping RCTPromiseRejectBlock) {
    let locale = Locale(identifier: language)
    speechRecognizer = SFSpeechRecognizer(locale: locale)
    guard let recognizer = speechRecognizer, recognizer.isAvailable else {
      reject("unavailable", "Speech recognizer not available for locale \(language)", nil)
      return
    }

    do {
      try startAudioSession()
      recognitionRequest = SFSpeechAudioBufferRecognitionRequest()
      guard let request = recognitionRequest else { return }

      request.shouldReportPartialResults = true
      request.requiresOnDeviceRecognition = true  // iOS 17+ on-device only

      // Custom vocabulary hint
      if !customVocabulary.isEmpty {
        request.contextualStrings = customVocabulary
      }

      sessionStartMs = Int64(Date().timeIntervalSince1970 * 1000)

      recognitionTask = recognizer.recognitionTask(with: request) { [weak self] result, error in
        guard let self = self else { return }

        if let error = error {
          self.sendEvent(withName: "onVoiceError", body: error.localizedDescription)
          return
        }

        guard let result = result else { return }

        let nowMs = Int64(Date().timeIntervalSince1970 * 1000)
        let latency = nowMs - self.sessionStartMs
        let stability: String = result.isFinal ? "final" : "partial"
        let text = result.bestTranscription.formattedString
        let confidence: Double? = result.isFinal ?
          Double(result.bestTranscription.segments.last?.confidence ?? 0) : nil

        let event: [String: Any] = [
          "text": text,
          "stability": stability,
          "timestamp_ms": nowMs,
          "latency_ms_from_audio_start": latency,
          "confidence": confidence as Any,
          "engine_metadata": ["engine": "apple_speech_transcriber", "is_final": result.isFinal]
        ]
        self.sendEvent(withName: "onTranscriptEvent", body: event)
      }

      let inputNode = audioEngine.inputNode
      let format = inputNode.outputFormat(forBus: 0)
      inputNode.installTap(onBus: 0, bufferSize: 1024, format: format) { [weak self] buffer, _ in
        guard let self = self, !self.isMuted else { return }
        self.recognitionRequest?.append(buffer)
      }

      audioEngine.prepare()
      try audioEngine.start()

      sendEvent(withName: "onVoiceStateChange", body: "listening")
      resolve(nil)
    } catch {
      reject("start_error", error.localizedDescription, error)
    }
  }

  @objc func stopListening(_ resolve: @escaping RCTPromiseResolveBlock,
                            rejecter reject: @escaping RCTPromiseRejectBlock) {
    audioEngine.stop()
    audioEngine.inputNode.removeTap(onBus: 0)
    recognitionRequest?.endAudio()
    recognitionTask?.cancel()
    recognitionRequest = nil
    recognitionTask = nil
    sendEvent(withName: "onVoiceStateChange", body: "stopped")
    resolve(nil)
  }

  @objc func mute() {
    isMuted = true
    sendEvent(withName: "onVoiceStateChange", body: "muted")
  }

  @objc func unmute() {
    isMuted = false
    sendEvent(withName: "onVoiceStateChange", body: "listening")
  }

  // MARK: - Private

  private func startAudioSession() throws {
    let session = AVAudioSession.sharedInstance()
    try session.setCategory(.playAndRecord,
                            mode: .measurement,
                            options: [.defaultToSpeaker, .allowBluetooth, .mixWithOthers])
    try session.setActive(true, options: .notifyOthersOnDeactivation)
  }
}
